juju/charm.py

```diff
@@ -82,15 +83,28 @@ def __init__(self, framework, key, metadata):
         super().__init__(framework, key)
         self.metadata = metadata
 
+        # These are a per-relation/storage group views of events, which are used for more programmatic access.
+        # For example: framework.observe(self.on.relation[db_rel].joined, self)
+        type(self.on).relation = {relation_name: SimpleNamespace() for relation_name in self.metadata.relations}
+        type(self.on).storage = {storage_name: SimpleNamespace() for storage_name in self.metadata.storage}
+
         for relation_name in self.metadata.relations:
             self.on.define_event(f'{relation_name}_relation_joined', RelationJoinedEvent)
             self.on.define_event(f'{relation_name}_relation_changed', RelationChangedEvent)
             self.on.define_event(f'{relation_name}_relation_departed', RelationDepartedEvent)
             self.on.define_event(f'{relation_name}_relation_broken', RelationBrokenEvent)
+            # Populate the group view for this relation.
+            self.on.relation[relation_name].joined = getattr(self.on, f'{relation_name}_relation_joined')
+            self.on.relation[relation_name].changed = getattr(self.on, f'{relation_name}_relation_changed')
+            self.on.relation[relation_name].departed = getattr(self.on, f'{relation_name}_relation_departed')
+            self.on.relation[relation_name].broken = getattr(self.on, f'{relation_name}_relation_broken')
 
         for storage_name in metadata.storage:
             self.on.define_event(f'{storage_name}_storage_attached', StorageAttachedEvent)
             self.on.define_event(f'{storage_name}_storage_detaching', StorageDetachingEvent)
+            # Populate the group view for this storage.
+            self.on.storage[storage_name].attached = getattr(self.on, f'{storage_name}_storage_attached')
```

Comment:
Hmmm.. that doesn't look entirely right. By getting the attribute we'll end up with a BoundEvent, which has a defined emitter on it. But we're assigning it to a property on a type. We may need something here which looks more like the current Event mechanism.